File: backend/integration/fixtures/fixtures.go
// Copyright (C) 2025 Austin Beattie
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

//go:build integration

// Package fixtures provides test data factories for integration tests.
package fixtures

import (
	"context"
	"database/sql"
	"fmt"
	"sync/atomic"
	"testing"
	"time"

	"github.com/octobud-hq/octobud/backend/internal/db"
)

// Counter for generating unique IDs
var counter int64

func nextID() int64 {
	return atomic.AddInt64(&counter, 1)
}

// RepositoryBuilder builds test repositories.
type RepositoryBuilder struct {
	githubID   int64
	name       string
	fullName   string
	ownerLogin string
}

// NewRepository creates a new repository builder with defaults.
func NewRepository() *RepositoryBuilder {
	id := nextID()
	return &RepositoryBuilder{
		githubID:   id,
		name:       fmt.Sprintf("test-repo-%d", id),
		fullName:   fmt.Sprintf("test-org/test-repo-%d", id),
		ownerLogin: "test-org",
	}
}

// WithGithubID sets the GitHub ID.
func (b *RepositoryBuilder) WithGithubID(id int64) *RepositoryBuilder {
	b.githubID = id
	return b
}

// WithName sets the repository name.
func (b *RepositoryBuilder) WithName(name string) *RepositoryBuilder {
	b.name = name
	return b
}

// WithFullName sets the full repository name.
func (b *RepositoryBuilder) WithFullName(fullName string) *RepositoryBuilder {
	b.fullName = fullName
	return b
}

// WithOwnerLogin sets the owner login.
func (b *RepositoryBuilder) WithOwnerLogin(ownerLogin string) *RepositoryBuilder {
	b.ownerLogin = ownerLogin
	return b
}

// Build creates the repository in the database.
func (b *RepositoryBuilder) Build(t *testing.T, ctx context.Context, store db.Store, userID string) db.Repository {
	t.Helper()

	repo, err := store.UpsertRepository(ctx, userID, db.UpsertRepositoryParams{
		GithubID:   sql.NullInt64{Int64: b.githubID, Valid: true},
		Name:       b.name,
		FullName:   b.fullName,
		OwnerLogin: sql.NullString{String: b.ownerLogin, Valid: true},
	})
	if err != nil {
		t.Fatalf("Failed to create repository: %v", err)
	}

	return repo
}

// NotificationBuilder builds test notifications.
type NotificationBuilder struct {
	githubID        string
	repositoryID    int64
	subjectType     string
	subjectTitle    string
	reason          string
	archived        bool
	isRead          bool
	muted           bool
	starred         bool
	filtered        bool
	snoozedUntil    sql.NullTime
	githubUpdatedAt sql.NullTime
}

// NewNotification creates a new notification builder with defaults.
func NewNotification(repositoryID int64) *NotificationBuilder {
	id := nextID()
	return &NotificationBuilder{
		githubID:        fmt.Sprintf("test-notif-%d", id),
		repositoryID:    repositoryID,
		subjectType:     "PullRequest",
		subjectTitle:    fmt.Sprintf("Test Notification %d", id),
		reason:          "subscribed",
		githubUpdatedAt: sql.NullTime{Time: time.Now().Add(-1 * time.Hour), Valid: true},
	}
}

// WithGithubID sets the GitHub ID.
func (b *NotificationBuilder) WithGithubID(id string) *NotificationBuilder {
	b.githubID = id
	return b
}

// WithSubjectType sets the subject type.
func (b *NotificationBuilder) WithSubjectType(t string) *NotificationBuilder {
	b.subjectType = t
	return b
}

// WithSubjectTitle sets the subject title.
func (b *NotificationBuilder) WithSubjectTitle(title string) *NotificationBuilder {
	b.subjectTitle = title
	return b
}

// WithReason sets the reason.
func (b *NotificationBuilder) WithReason(reason string) *NotificationBuilder {
	b.reason = reason
	return b
}

// WithArchived sets the archived flag.
func (b *NotificationBuilder) WithArchived(archived bool) *NotificationBuilder {
	b.archived = archived
	return b
}

// WithIsRead sets the is_read flag.
func (b *NotificationBuilder) WithIsRead(isRead bool) *NotificationBuilder {
	b.isRead = isRead
	return b
}

// WithMuted sets the muted flag.
func (b *NotificationBuilder) WithMuted(muted bool) *NotificationBuilder {
	b.muted = muted
	return b
}

// WithStarred sets the starred flag.
func (b *NotificationBuilder) WithStarred(starred bool) *NotificationBuilder {
	b.starred = starred
	return b
}

// WithFiltered sets the filtered flag.
func (b *NotificationBuilder) WithFiltered(filtered bool) *NotificationBuilder {
	b.filtered = filtered
	return b
}

// WithSnoozedUntil sets the snooze time.
func (b *NotificationBuilder) WithSnoozedUntil(t time.Time) *NotificationBuilder {
	b.snoozedUntil = sql.NullTime{Time: t, Valid: true}
	return b
}

// WithGithubUpdatedAt sets the github_updated_at time.
func (b *NotificationBuilder) WithGithubUpdatedAt(t time.Time) *NotificationBuilder {
	b.githubUpdatedAt = sql.NullTime{Time: t, Valid: true}
	return b
}

// Build creates the notification in the database.
func (b *NotificationBuilder) Build(t *testing.T, ctx context.Context, store db.Store, userID string) db.Notification {
	t.Helper()

	// Create notification via upsert
	notif, err := store.UpsertNotification(ctx, userID, db.UpsertNotificationParams{
		GithubID:        b.githubID,
		RepositoryID:    b.repositoryID,
		SubjectType:     b.subjectType,
		SubjectTitle:    b.subjectTitle,
		Reason:          sql.NullString{String: b.reason, Valid: true},
		GithubUpdatedAt: b.githubUpdatedAt,
	})
	if err != nil {
		t.Fatalf("Failed to create notification: %v", err)
	}

	// Apply additional states if needed
	if b.archived {
		notif, err = store.ArchiveNotification(ctx, userID, b.githubID)
		if err != nil {
			t.Fatalf("Failed to archive notification: %v", err)
		}
	}

	if b.isRead {
		notif, err = store.MarkNotificationRead(ctx, userID, b.githubID)
		if err != nil {
			t.Fatalf("Failed to mark notification read: %v", err)
		}
	}

	if b.muted {
		notif, err = store.MuteNotification(ctx, userID, b.githubID)
		if err != nil {
			t.Fatalf("Failed to mute notification: %v", err)
		}
	}

	if b.starred {
		notif, err = store.StarNotification(ctx, userID, b.githubID)
		if err != nil {
			t.Fatalf("Failed to star notification: %v", err)
		}
	}

	if b.filtered {
		notif, err = store.MarkNotificationFiltered(ctx, userID, b.githubID)
		if err != nil {
			t.Fatalf("Failed to filter notification: %v", err)
		}
	}

	if b.snoozedUntil.Valid {
		notif, err = store.SnoozeNotification(ctx, userID, db.SnoozeNotificationParams{
			GithubID:     b.githubID,
			SnoozedUntil: b.snoozedUntil,
		})
		if err != nil {
			t.Fatalf("Failed to snooze notification: %v", err)
		}
	}

	return notif
}

// TagBuilder builds test tags.
type TagBuilder struct {
	name        string
	slug        string
	color       string
	description string
}

// NewTag creates a new tag builder with defaults.
func NewTag() *TagBuilder {
	id := nextID()
	return &TagBuilder{
		name:  fmt.Sprintf("Test Tag %d", id),
		slug:  fmt.Sprintf("test-tag-%d", id),
		color: "#FF5733",
	}
}

// WithName sets the tag name.
func (b *TagBuilder) WithName(name string) *TagBuilder {
	b.name = name
	return b
}

// WithSlug sets the tag slug.
func (b *TagBuilder) WithSlug(slug string) *TagBuilder {
	b.slug = slug
	return b
}

// WithColor sets the tag color.
func (b *TagBuilder) WithColor(color string) *TagBuilder {
	b.color = color
	return b
}

// WithDescription sets the tag description.
func (b *TagBuilder) WithDescription(desc string) *TagBuilder {
	b.description = desc
	return b
}

// Build creates the tag in the database.
func (b *TagBuilder) Build(t *testing.T, ctx context.Context, store db.Store, userID string) db.Tag {
	t.Helper()

	tag, err := store.UpsertTag(ctx, userID, db.UpsertTagParams{
		Name:        b.name,
		Slug:        b.slug,
		Color:       sql.NullString{String: b.color, Valid: true},
		Description: sql.NullString{String: b.description, Valid: b.description != ""},
	})
	if err != nil {
		t.Fatalf("Failed to create tag: %v", err)
	}

	return tag
}

// AssignTag assigns a tag to a notification.
func AssignTag(t *testing.T, ctx context.Context, store db.Store, userID string, tagID string, notificationID int64) {
	t.Helper()

	_, err := store.AssignTagToEntity(ctx, userID, db.AssignTagToEntityParams{
		TagID:      tagID,
		EntityType: "notification",
		EntityID:   notificationID,
	})
	if err != nil {
		t.Fatalf("Failed to assign tag: %v", err)
	}

	// Note: SQLite uses a junction table, so no need to update a denormalized tag_ids column
}

File: extension/dev/stub-server/fixtures.mjs
// Copyright (C) 2025 Austin Beattie
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

/**
 * Seed data for the stub server, shaped exactly like
 * backend/internal/models.Notification's JSON so the vendored
 * `fromBackendNotification` mapper works unchanged.
 *
 * The set is chosen to cover every visual branch NotificationRow can take:
 * each subject type and state, read/unread, starred, muted, archived, snoozed,
 * filtered, tagged (including more than the three the row shows inline), and a
 * title long enough to wrap in a narrow panel.
 */

const MINUTE = 60 * 1000;
const HOUR = 60 * MINUTE;
const DAY = 24 * HOUR;

const ago = (ms) => new Date(Date.now() - ms).toISOString();
const ahead = (ms) => new Date(Date.now() + ms).toISOString();

export const REPOSITORIES = {
	octobud: {
		id: 1,
		githubId: 900001,
		name: "octobud",
		fullName: "octobud-hq/octobud",
		ownerLogin: "octobud-hq",
		htmlUrl: "https://github.com/octobud-hq/octobud",
		private: false,
	},
	habitflow: {
		id: 2,
		githubId: 900002,
		name: "habitflow",
		fullName: "addonovan/habitflow",
		ownerLogin: "addonovan",
		htmlUrl: "https://github.com/addonovan/habitflow",
		private: true,
	},
	svelte: {
		id: 3,
		githubId: 900003,
		name: "svelte",
		fullName: "sveltejs/svelte",
		ownerLogin: "sveltejs",
		htmlUrl: "https://github.com/sveltejs/svelte",
		private: false,
	},
};

export const TAGS = [
	{ id: "tag-1", name: "urgent", slug: "urgent", color: "#ef4444", unreadCount: 2 },
	{ id: "tag-2", name: "backend", slug: "backend", color: "#3b82f6", unreadCount: 3 },
	{ id: "tag-3", name: "design", slug: "design", color: "#a855f7", unreadCount: 1 },
	{ id: "tag-4", name: "chore", slug: "chore", color: "#64748b", unreadCount: 0 },
	{ id: "tag-5", name: "flaky", slug: "flaky", color: "#f59e0b", unreadCount: 1 },
];

const tagBySlug = (slug) => TAGS.find((tag) => tag.slug === slug);

let nextId = 1;

function build({
	repo,
	subjectType,
	title,
	number,
	reason,
	author,
	state,
	merged = false,
	draft = false,
	isRead = false,
	starred = false,
	muted = false,
	archived = false,
	filtered = false,
	snoozedUntil = null,
	tags = [],
	updatedAt,
}) {
	const id = nextId++;
	const repository = REPOSITORIES[repo];
	const pathSegment = subjectType === "Issue" ? "issues" : "pull";
	const htmlUrl = number ? `${repository.htmlUrl}/${pathSegment}/${number}` : repository.htmlUrl;

	return {
		id,
		githubId: `stub-${String(id).padStart(4, "0")}`,
		repositoryId: repository.id,
		subjectType,
		subjectTitle: title,
		subjectUrl: number
			? `https://api.github.com/repos/${repository.fullName}/${pathSegment === "pull" ? "pulls" : "issues"}/${number}`
			: null,
		reason,
		archived,
		isRead,
		muted,
		starred,
		filtered,
		snoozedUntil,
		snoozedAt: snoozedUntil ? ago(2 * HOUR) : null,
		effectiveSortDate: updatedAt,
		githubUnread: !isRead,
		githubUpdatedAt: updatedAt,
		githubUrl: `https://api.github.com/notifications/threads/${id}`,
		importedAt: updatedAt,
		repository,
		subjectNumber: number ?? null,
		subjectState: state ?? null,
		subjectMerged: merged,
		subjectDraft: draft,
		authorLogin: author,
		tags: tags.map(tagBySlug).filter(Boolean),
		labels: [],
		assignees: [],
		reviewers: [],
		teamReviewers: [],
		subjectRaw: {
			title,
			html_url: htmlUrl,
			number: number ?? undefined,
			state: state ?? undefined,
			draft,
			merged,
			user: { login: author },
			updated_at: updatedAt,
		},
		subjectFetchedAt: updatedAt,
	};
}

export function createNotifications() {
	nextId = 1;
	return [
		build({
			repo: "octobud",
			subjectType: "PullRequest",
			title: "Add a browser side panel extension",
			number: 121,
			reason: "review_requested",
			author: "addonovan",
			state: "open",
			updatedAt: ago(18 * MINUTE),
			tags: ["urgent", "backend"],
		}),
		build({
			repo: "octobud",
			subjectType: "PullRequest",
			title: "Fix the flaky notification sync integration test that intermittently fails on CI",
			number: 119,
			reason: "author",
			author: "addonovan",
			state: "open",
			updatedAt: ago(2 * HOUR),
			tags: ["flaky"],
		}),
		build({
			repo: "octobud",
			subjectType: "PullRequest",
			title: "Bump modernc.org/sqlite from 1.50.0 to 1.50.1",
			number: 118,
			reason: "subscribed",
			author: "dependabot",
			state: "closed",
			merged: true,
			isRead: true,
			updatedAt: ago(6 * HOUR),
			tags: ["chore"],
		}),
		build({
			repo: "octobud",
			subjectType: "PullRequest",
			title: "Draft: rework the query parser error messages",
			number: 117,
			reason: "mention",
			author: "kbeattie",
			state: "open",
			draft: true,
			updatedAt: ago(9 * HOUR),
		}),
		build({
			repo: "octobud",
			subjectType: "Issue",
			title: "Views with an empty query should fall back to inbox semantics",
			number: 116,
			reason: "mention",
			author: "octocat",
			state: "open",
			updatedAt: ago(1 * DAY),
			tags: ["backend"],
		}),
		build({
			repo: "habitflow",
			subjectType: "Issue",
			title: "Streak resets an hour early in non-UTC time zones",
			number: 42,
			reason: "assign",
			author: "addonovan",
			state: "closed",
			isRead: true,
			updatedAt: ago(2 * DAY),
		}),
		build({
			repo: "habitflow",
			subjectType: "PullRequest",
			title: "Weekly review charts",
			number: 91,
			reason: "review_requested",
			author: "addonovan",
			state: "open",
			starred: true,
			updatedAt: ago(40 * MINUTE),
			tags: ["design"],
		}),
		build({
			repo: "habitflow",
			subjectType: "PullRequest",
			title: "Auto-miss overdue occurrences via Solid Queue",
			number: 88,
			reason: "subscribed",
			author: "addonovan",
			state: "closed",
			merged: true,
			archived: true,
			isRead: true,
			updatedAt: ago(3 * DAY),
		}),
		build({
			repo: "habitflow",
			subjectType: "PullRequest",
			title: "Push notification egress proxy",
			number: 100,
			reason: "author",
			author: "addonovan",
			state: "open",
			snoozedUntil: ahead(20 * HOUR),
			updatedAt: ago(5 * HOUR),
			tags: ["backend", "urgent"],
		}),
		build({
			repo: "svelte",
			subjectType: "Release",
			title: "svelte@5.55.7",
			number: null,
			reason: "subscribed",
			author: "sveltejs",
			updatedAt: ago(11 * HOUR),
		}),
		build({
			repo: "svelte",
			subjectType: "Discussion",
			title: "RFC: attachment syntax for actions",
			number: 14002,
			reason: "subscribed",
			author: "rich-harris",
			isRead: true,
			updatedAt: ago(4 * DAY),
		}),
		build({
			repo: "svelte",
			subjectType: "Issue",
			title: "Legacy component API deprecation warning is too noisy",
			number: 13990,
			reason: "subscribed",
			author: "octocat",
			state: "open",
			muted: true,
			updatedAt: ago(6 * DAY),
		}),
		build({
			repo: "octobud",
			subjectType: "CheckSuite",
			title: "CI workflow failed on main",
			number: null,
			reason: "ci_activity",
			author: "github-actions",
			state: "failure",
			updatedAt: ago(30 * MINUTE),
			tags: ["urgent"],
		}),
		build({
			repo: "octobud",
			subjectType: "PullRequest",
			title: "Tidy up the sidebar collapse animation",
			number: 112,
			reason: "team_mention",
			author: "kbeattie",
			state: "open",
			filtered: true,
			updatedAt: ago(3 * DAY),
			tags: ["design", "chore"],
		}),
		build({
			repo: "octobud",
			subjectType: "PullRequest",
			title: "Query engine: support parenthesised OR groups",
			number: 108,
			reason: "review_requested",
			author: "octocat",
			state: "open",
			isRead: true,
			updatedAt: ago(5 * DAY),
			tags: ["backend", "urgent", "design", "chore"],
		}),
		build({
			repo: "habitflow",
			subjectType: "Commit",
			title: "Fix N+1 in the agenda serializer",
			number: null,
			reason: "comment",
			author: "addonovan",
			updatedAt: ago(7 * HOUR),
		}),
	];
}

/**
 * Custom views alongside the five the backend synthesizes. Queries use the same
 * syntax the real engine accepts, within the stub evaluator's subset.
 */
export const SYSTEM_VIEWS = [
	{ slug: "inbox", name: "Inbox", icon: "inbox", query: "in:inbox", isDefault: true },
	{ slug: "everything", name: "Everything", icon: "infinity", query: "in:anywhere" },
	{ slug: "starred", name: "Starred", icon: "star", query: "is:starred in:anywhere" },
	{ slug: "snoozed", name: "Snoozed", icon: "snooze", query: "in:snoozed" },
	{ slug: "archive", name: "Archive", icon: "archive", query: "in:archive" },
];

export const USER_VIEWS = [
	{
		slug: "review-requests",
		name: "Needs my review",
		icon: "👀",
		description: "Open PRs waiting on me",
		query: "reason:review_requested is:unread",
	},
	{
		slug: "my-prs",
		name: "My pull requests",
		icon: "✨",
		query: "type:PullRequest author:addonovan",
	},
	{
		slug: "urgent",
		name: "Urgent",
		icon: "🔥",
		description: "Anything tagged urgent",
		query: "tags:urgent in:anywhere",
	},
	{
		slug: "ci",
		name: "CI activity",
		icon: "🤖",
		query: "reason:ci_activity in:anywhere",
	},
];
